### PR TITLE
Surface mesh, a generic representation for polygonal surfaces

### PR DESCRIPTION
**Pull request by JacquesOlivierLachaud**
# PR Description

This PR provides classes to represent and process generic polygonal faces, whether manifold or non manifold.

# Checklist

- [x] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [x] Doxygen documentation of the code completed (classes, methods, types, members...)
- [x] Documentation module page added or updated.
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [x] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [x] All continuous integration tests pass (Travis & appveyor)

### CODE REVIEW

File: src/DGtal/shapes/SurfaceMesh.h [hunk 1]
```diff
@@ -0,0 +1,750 @@
...
+    bool setVertexNormals( RealVectorIterator itN, RealVectorIterator itNEnd );
+
+    /// Given a range of real vectors, sets the normals of every
+    /// face to the given vectors.
+    template <typename RealVectorIterator>
+    bool setFaceNormals( RealVectorIterator itN, RealVectorIterator itNEnd );
+
+    /// Uses the positions of vertices to compute a normal vector to
+    /// each face of the mesh. It computes the barycenter,
+    /// triangulates implicitly the face to build the normal vector.
```
Comment (kerautret) [inline src/DGtal/shapes/SurfaceMesh.h:209]:
```suggestion
    /// triangulates implicitly the face to build the normal vector from the average of implicit triangle normals.
```

File: src/DGtal/shapes/SurfaceMesh.ih [hunk 2]
```diff
@@ -0,0 +1,858 @@
...
+                trace.error() << "bad edge " << vi << " " << vn << std::endl;
+                continue;
+              }
+              Scalar eweight = edgeInclusionRatio( p, r, e_ij );
+              if ( eweight > 0.0 )
+                result_e.push_back( std::make_pair( e_ij, eweight ) );
+            }
+        }
+    }
+  // trace.info() << "#Marked=" << marked.size()
```
Comment (kerautret) [inline src/DGtal/shapes/SurfaceMesh.ih:554]:
```suggestion
```

File: src/DGtal/shapes/SurfaceMesh.ih [hunk 3]
```diff
@@ -0,0 +1,858 @@
...
+  //                << " #Result=" << result.size() << std::endl;
```
Comment (kerautret) [inline src/DGtal/shapes/SurfaceMesh.ih:555]:
```suggestion
```

File: src/DGtal/shapes/SurfaceMeshHelper.ih [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+  const auto fv = [n] (Size i, Size j) { return i * n + j + (Size)1; };
+  for ( Size i = 0; i < m; ++i )
+    for ( Size j = 0; j < n; ++j )
+      {
+        const Scalar theta = ( 2.0 * M_PI * (double) j ) / (double) n;
+        const Scalar   phi = ( M_PI * (double) (i+1) ) / (double) (m+1) - 0.5 * M_PI;
+        p[ fv( i, j ) ] = center + radius * RealPoint( cos( theta ) * cos( phi ),
+                                                       sin( theta ) * cos( phi ),
+                                                       sin( phi ) );
+        // trace.info() << fv( i, j ) << "=(" << i << "," << j << ")"
```
Comment (kerautret) [inline src/DGtal/shapes/SurfaceMeshHelper.ih:72]: clean ? (perhaps we could add later a trace.debug())

File: src/DGtal/shapes/SurfaceMeshHelper.ih [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+    for ( Size j = 0; j < n; ++j )
+      {
+        const Scalar theta = ( i % 2 == 0 )
+          ? ( 2.0 * M_PI * (double) j ) / (double) n
+          : ( 2.0 * M_PI * ( 0.5 + (double) j ) ) / (double) n;
+        const Scalar     h = height * ( (double) i / (double) (m-1) - 0.5 );
+        p[ fv( i, j ) ] = center + RealPoint( radius * cos( theta ),
+                                              radius * sin( theta ),
+                                              h );
+        // trace.info() << fv( i, j ) << "=(" << i << "," << j << ")"
```
Comment (kerautret) [inline src/DGtal/shapes/SurfaceMeshHelper.ih:269]: same here

File: src/DGtal/shapes/SurfaceMeshHelper.ih [hunk 6]
```diff
@@ -0,0 +1,643 @@
...
+    for ( Size j = 0; j < n; ++j )
+      {
+        const Scalar theta = ( 2.0 * M_PI * (double) i ) / (double) m;
+        const Scalar   phi = ( 2.0 * M_PI * (double) j ) / (double) n
+          + twist * 2.0 * M_PI * (double) i / ( double ) m;
+        p[ fv( i, j ) ] = center
+          + RealPoint( ( big_radius + small_radius * cos( phi ) ) * cos( theta ),
+                       ( big_radius + small_radius * cos( phi ) ) * sin( theta ),
+                       small_radius * sin( phi ) );
+        // trace.info() << fv( i, j ) << "=(" << i << "," << j << ")"
```
Comment (kerautret) [inline src/DGtal/shapes/SurfaceMeshHelper.ih:445]: same

File: src/DGtal/shapes/SurfaceMesh.h [hunk 7]
```diff
@@ -0,0 +1,773 @@
...
+  // template class SurfaceMesh
+  /**
+     Description of template class 'SurfaceMesh' <p> \brief Aim:
+     Represents a mesh as faces and a list of vertices. Vertices may
+     be shared among faces but no correct topology is
+     required. However, you also have methods to navigate between
+     neighbor vertices, faces, etc. The mesh can be equipped with
+     normals at faces and/or vertices.
+
+     We sum up below the possible classes for representing meshes:
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMesh.h:62]: Maybe this should be copied (or moved elsewhere). 
Eg:
 
* on packageShapes create a new subitem level grouping "Manipulating Simple Mesh (Bertrand Kerautret)", 
"Half-edge data structure, triangulated surfaces and polygonal surfaces (Jacques-Olivier Lachaud)" and "Surface mesh data structure for representing manifold or non-manifold polygonal surfaces in R3 (Jacques-Olivier Lachaud)
*  Just copie this paragraph to a very short module page heading the new subitems list

Do you see what I mean ? I have the feeling that this list should be top level and preliminary to the dedicated module page

File: src/DGtal/shapes/SurfaceMesh.h [hunk 8]
```diff
@@ -0,0 +1,773 @@
...
+    /// The type for counting elements.
+    typedef std::size_t                             Size;
+    /// The type used for numbering vertices and faces
+    typedef std::size_t                             Index;
+    typedef Index                                   Face;
+    typedef Index                                   Edge;
+    typedef Index                                   Vertex;
+    typedef std::pair< Edge, Scalar >               WeightedEdge;
+    typedef std::pair< Face, Scalar >               WeightedFace;
+    /// The type that defines a range of vertices
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMesh.h:111]:
```suggestion
    /// The type that defines a range of vertices (e.g. to define faces)
```

File: src/DGtal/shapes/SurfaceMesh.h [hunk 9]
```diff
@@ -0,0 +1,773 @@
...
+    typedef std::vector< Vertex >                   Vertices;
+    typedef std::vector< Vertex >                   VertexRange;
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMesh.h:113]: I would only keep one of these two types

File: src/DGtal/shapes/SurfaceMesh.h [hunk 10]
```diff
@@ -0,0 +1,773 @@
...
+    /// The type that defines a range of faces
+    typedef std::vector< Edge >                     Edges;
+    typedef std::vector< Edge >                     EdgeRange;
+    typedef std::vector< WeightedEdge >             WeightedEdges;
+    typedef std::vector< WeightedEdge >             WeightedEdgeRange;
+    typedef std::vector< Face >                     Faces;
+    typedef std::vector< Face >                     FaceRange;
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMesh.h:120]: Same here

Comment (dcoeurjo) [reply src/DGtal/shapes/SurfaceMesh.h:120]: The convention "SomeTypes" vs. "SomeTypeRange" may not be necessary and could confuse the reader

File: src/DGtal/shapes/SurfaceMesh.h [hunk 11]
```diff
@@ -0,0 +1,773 @@
...
+    SurfaceMesh( Self&& other ) = default;
+    /// Default assignment constructor.
+    /// @param other the object to clone
+    /// @return a reference to 'this'.
+    Self& operator=( const Self& other ) = default;
+
+    /// Builds a mesh from vertex positions and polygonal faces.
+    ///
+    /// @tparam RealPointIterator any forward iterator on RealPoint.
+    /// @tparam VerticesIterator any forward iterator on a range of vertices.
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMesh.h:167]:
```suggestion
    /// @tparam VerticesIterator any forward iterator on the range of vertices defining a face.
```

File: src/DGtal/shapes/SurfaceMesh.h [hunk 12]
```diff
@@ -0,0 +1,773 @@
...
+    void computeFaceNormalsFromVertexNormals();
+
+    /// Uses the normals associated with faces to compute a normal
+    /// vector to each vertex of the mesh. It simply averages the
+    /// normals of every incident face.
+    void computeVertexNormalsFromFaceNormals();
+    
+    /// Uses the normals associated with faces to compute a normal
+    /// vector to each vertex of the mesh. It uses the weights
+    /// proposed by [Max, 1999] for combining face information into
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMesh.h:247]: Explicit bibtex ref?

File: src/DGtal/shapes/SurfaceMesh.h [hunk 13]
```diff
@@ -0,0 +1,773 @@
...
+#include "DGtal/base/IntegerSequenceIterator.h"
+#include "DGtal/helpers/StdDefs.h"
+
+namespace DGtal
+{
+  /////////////////////////////////////////////////////////////////////////////
+  // template class SurfaceMesh
+  /**
+     Description of template class 'SurfaceMesh' <p> \brief Aim:
+     Represents a mesh as faces and a list of vertices. Vertices may
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMesh.h:56]:
```suggestion
     Represents an embedded mesh as faces and a list of vertices. Vertices may
```
?

File: src/DGtal/shapes/SurfaceMesh.h [hunk 14]
```diff
@@ -0,0 +1,773 @@
...
+    Scalar faceArea( Index f ) const;
+
+    /// @param v any valid vertex index.
+    /// @return the Max's weights for each incident face to \a v, in the same order as `myIncidentFaces[ v ]`.
+    /// @note Used in computeVertexNormalsFromFaceNormalsWithMaxWeights
+    Scalars getMaxWeights( Index v ) const;
+    
+    /// Given a ball of radius \a r centered on the centroid of face
+    /// \a f, return the faces having an non empty intersection with
+    /// this ball, each one weighted by its ratio of inclusion.
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMesh.h:625]:
```suggestion
    /// this ball, each one weighted by its ratio of inclusion: for edges, the ratio between the length of the part of the edge included in the ball and the total edge length. For faces, the ratio between the @e approximated @e area of the face part inside the ball and the face area).
```

def "inclusion" missing

File: src/DGtal/shapes/SurfaceMesh.h [hunk 15]
```diff
@@ -0,0 +1,773 @@
...
+    ///
+    /// @param r the radius of the ball.
+    /// @param f the face where the ball is centered.
+    ///
+    /// @return the range of faces having an non empty intersection
+    /// with this ball, each one weighted by its ratio of inclusion.
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMesh.h:631]: cut paste the def of "inclusion" ?

File: src/DGtal/shapes/SurfaceMesh.h [hunk 16]
```diff
@@ -0,0 +1,773 @@
...
+    WeightedFaces
+    computeFacesInclusionsInBall( Scalar r, Index f ) const;
+    
+    /// Given a ball of radius \a r centered on the centroid of face
+    /// \a f, return the vertices/edges/faces having an non empty intersection with
+    /// this ball, each edge/face weighted by its ratio of inclusion.
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMesh.h:637]: same here

File: src/DGtal/shapes/SurfaceMesh.h (hunk 15 above)
Comment (dcoeurjo) [reply src/DGtal/shapes/SurfaceMesh.h:631]: or refer to the first function

File: src/DGtal/shapes/SurfaceMesh.h (hunk 14 above)
Comment (dcoeurjo) [reply src/DGtal/shapes/SurfaceMesh.h:625]: +some details on the approx?

File: src/DGtal/shapes/SurfaceMeshHelper.h [hunk 17]
```diff
@@ -0,0 +1,454 @@
...
+    /// @param[in] n the number of longitudes, minimum is 3.
+    ///
+    /// @param[in] normals specifies built-time normals of the output
+    /// mesh: NO_NORMALS no normals, VERTEX_NORMALS the normals at
+    /// each vertex, FACE_NORMALS the normals at each face.
+    ///
+    /// @return the corresponding surface mesh.
+    static
+    SurfaceMesh
+    makeSphere( Scalar radius, RealPoint center,
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMeshHelper.h:106]:
```suggestion
    makeSphere( const Scalar radius, const RealPoint &center,
```

File: src/DGtal/shapes/SurfaceMeshHelper.h [hunk 18]
```diff
@@ -0,0 +1,454 @@
...
+                Size m, Size n, Normals normals );
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMeshHelper.h:107]:
```suggestion
                const Size m, const Size n, Normals normals );
```

File: src/DGtal/shapes/SurfaceMeshHelper.h [hunk 19]
```diff
@@ -0,0 +1,454 @@
...
+    /// longitudes.
+    ///
+    /// @param[in] radius the radius of the sphere.
+    /// @param[in] m the number of latitudes (poles excepted), minimum is 1.
+    /// @param[in] n the number of longitudes, minimum is 3.
+    ///
+    /// @return the mean curvature at each vertex.
+    static
+    Scalars
+    sphereMeanCurvatures( Scalar radius, Size m, Size n );
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMeshHelper.h:120]: maybe `const` missing everywhere

File: src/DGtal/shapes/SurfaceMeshHelper.h [hunk 20]
```diff
@@ -0,0 +1,454 @@
...
+    typedef std::vector<RealVector>                 RealVectors;
+    typedef typename SurfaceMesh::Size           Size;
+    typedef typename SurfaceMesh::Index          Index;
+    typedef typename SurfaceMesh::Vertices       Vertices;
+    typedef typename SurfaceMesh::Faces          Faces;
+
+    /// Enum type for specifying if the mesh should have after
+    /// construction: NO_NORMALS no normals, VERTEX_NORMALS the
+    /// normals at each vertex, FACE_NORMALS the normals at each face.
+    enum class Normals { NO_NORMALS, VERTEX_NORMALS, FACE_NORMALS };
```
Comment (dcoeurjo) [inline src/DGtal/shapes/SurfaceMeshHelper.h:83]: I was confused between Normals as normal container (aka std::vector<RealVector> for instance, following previous convention) and this tag..

I would rename this as "NormalType/Tag" or something like that

File: src/DGtal/shapes/doc/moduleSurfaceMesh.dox [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+/*!
+@page moduleSurfaceMesh Surface mesh data structure for representing manifold or non-manifold polygonal surfaces in R3
+@author Jacques-Olivier Lachaud
+
+@since 1.1
+
+Part of \ref packageShapes
+  
+This part of the manual describes how to represent and manipulate
+generic polygonal surfaces embedded in \f$ \mathbb{R}^3 \f$. The class
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleSurfaceMesh.dox:29]: If we have a quick "Mesh types / variations" module page (see above), add ref to it here

File: src/DGtal/shapes/doc/moduleSurfaceMesh.dox [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+
+- by using classical predefined surfaces (see SurfaceMeshHelper::makeSphere, SurfaceMeshHelper::makeTorus, SurfaceMeshHelper::makeLantern).
+
+@snippet examples/shapes/exampleSurfaceMesh.cpp exampleSurfaceMesh-make-torus
+
+@image html exampleSurfaceMesh.jpg "Creating surface meshes from OBJ file, by specifying vertex/face information or predefined shapes"
+
+\section SurfMesh_sec2 Topological relations within a surface mesh
+
+All topological relations are precomputed as arrays in SurfaceMesh
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleSurfaceMesh.dox:87]:
```suggestion
All topological relations are precomputed as static arrays in SurfaceMesh
```

File: src/DGtal/shapes/doc/moduleSurfaceMesh.dox [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+class (which is thus not adapted to dynamic topological updates). You
+may access the number of cells with SurfaceMesh::nbVertices,
+SurfaceMesh::nbEdges, SurfaceMesh::nbFaces. Note that edge indices
+corresponds to pairs of vertices (i,j) with i<j.
+
+You may ask for each vertex \a v:
+- its vector of incident faces with SurfaceMesh::incidentFaces (order not significant)
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleSurfaceMesh.dox:94]:
```suggestion
- its range of incident faces with SurfaceMesh::incidentFaces (order not significant)
```
?

File: src/DGtal/shapes/doc/moduleSurfaceMesh.dox [hunk 24]
```diff
@@ -0,0 +1,237 @@
...
+- its vector of neighbor vertices with SurfaceMesh::neighborVertices (order not significant)
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleSurfaceMesh.dox:95]:
```suggestion
- its range of neighbor vertices with SurfaceMesh::neighborVertices (order not significant)
```

File: src/DGtal/shapes/doc/moduleSurfaceMesh.dox [hunk 25]
```diff
@@ -0,0 +1,237 @@
...
+
+You may ask for each face \a f:
+- its vector of incident vertices with SurfaceMesh::incidentVertices (the order is significant: to each consecutive pair of vertices, there is a corresponding edge).
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleSurfaceMesh.dox:98]:
```suggestion
- its range of incident vertices with SurfaceMesh::incidentVertices (the order is significant: to each consecutive pair of vertices, there is a corresponding edge).
```

File: src/DGtal/shapes/doc/moduleSurfaceMesh.dox [hunk 26]
```diff
@@ -0,0 +1,237 @@
...
+- its vector of neighbor faces with SurfaceMesh::neighborFaces (order not significant)
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleSurfaceMesh.dox:99]:
```suggestion
- its range of neighbor faces with SurfaceMesh::neighborFaces (order not significant)
```

File: src/DGtal/shapes/doc/moduleSurfaceMesh.dox [hunk 27]
```diff
@@ -0,0 +1,237 @@
...
+
+You may create an edge index with SurfaceMesh::makeEdge. If the two
+vertices (i,j) do not form an edge, then the returned index is
+SurfaceMesh::nbEdges. Note that calling `makeEdge(i,j)` or
+`makeEdge(j,i)` returns always the same index, whether valid or
+invalid.
+
+You may ask for each edge \a e:
+- its two incident vertices with SurfaceMesh::incidentVertices, as pair (i,j) with (i<j).
+- its vector of bordering faces with SurfaceMesh::edgeFaces (they can be incident clockwise or counterclockwise)
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleSurfaceMesh.dox:109]:
```suggestion
- its range of bordering faces with SurfaceMesh::edgeFaces (they can be incident clockwise or counterclockwise)
```

File: src/DGtal/shapes/doc/moduleSurfaceMesh.dox [hunk 28]
```diff
@@ -0,0 +1,237 @@
...
+- its vector of left bordering faces with SurfaceMesh::edgeLeftFaces (a face to its left, being defined ccw, means that the face is some `(..., i, j, ... )`)
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleSurfaceMesh.dox:110]:
```suggestion
- its range of left bordering faces with SurfaceMesh::edgeLeftFaces (a face to its left, being defined ccw, means that the face is some `(..., i, j, ... )`)
```

File: src/DGtal/shapes/doc/moduleSurfaceMesh.dox [hunk 29]
```diff
@@ -0,0 +1,237 @@
...
+- its vector of right bordering faces with SurfaceMesh::edgeRightFaces (a face to its left, being defined ccw, means that the face is some `(..., j, i, ... )`)
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleSurfaceMesh.dox:111]:
```suggestion
- its range of right bordering faces with SurfaceMesh::edgeRightFaces (a face to its left, being defined ccw, means that the face is some `(..., j, i, ... )`)
```

File: src/DGtal/shapes/doc/moduleSurfaceMesh.dox [hunk 30]
```diff
@@ -0,0 +1,237 @@
...
+- the centroid of a face with SurfaceMesh::faceCentroid
+- the centroid of an edge with SurfaceMesh::edgeCentroid
+- the average distance between the centroid of face and its vertices
+  with SurfaceMesh::localWindow
+- the classical Max's weights for each incident face to a vertex with
+  SurfaceMesh::getMaxWeights (see
+  SurfaceMesh::computeVertexNormalsFromFaceNormalsWithMaxWeights)
+- the vector of weighted faces comprising all the faces included in
+  some ball with SurfaceMesh::computeFacesInclusionsInBall, where the
+  weights represent the inclusion ratio.
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleSurfaceMesh.dox:205]: +def (cut/paste from the class def)?

File: src/DGtal/shapes/doc/packageShapes.dox [hunk 31]
```diff
@@ -53,14 +53,17 @@ make concrete such objects in a given grid step grid.
 
 Since 0.9.4, there is a module that provides classes and services to
 represent combinatorial surfaces, based on an half-edge data
-structure.
+structure. Since 1.1, there is a module that provides classes and
+services to represent generic polygonal surfaces, either manifold or
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/packageShapes.dox:57]:
```suggestion
services to represent generic embedded polygonal surfaces, either manifold or
```

**Review by dcoeurjo — CHANGES_REQUESTED** [review]
Thanks for this huge PR 
Many comments (with github suggestion) and a doc issue to clarify the positioning wrt to the other classes.

**Comment by kerautret** [comment]
@JacquesOlivierLachaud I also make a small PR on your branch with an add if you miss it (else no pb I will PR on master later)

**Comment by kerautret** [comment]
here: https://github.com/JacquesOlivierLachaud/DGtal/pulls

File: examples/shapes/exampleSurfaceMesh.cpp [hunk 32]
```diff
@@ -0,0 +1,163 @@
...
+ */
+
+
+/**
+   This snippet shows how to identify and display digital fully
+   subconvex sets of a grid curve form its tangent bundle.
+
+   @see \ref moduleSurfaceMesh
+
+   \image html grid-curve-dig-convexity.png "Extraction of all subconvex triangles to the digital curve."
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply examples/shapes/exampleSurfaceMesh.cpp:37]: done

File: src/DGtal/io/writers/SurfaceMeshWriter.h [hunk 33]
```diff
@@ -0,0 +1,214 @@
...
+    /// MTL file) and associate color information.
+    ///
+    /// @param[in] objfile the name of the OBJ file (like "bunny" or "bunny.obj").
+    /// @param[in] smesh the surface mesh.
+    ///
+    /// @param[in] diffuse_colors either empty or a vector containing
+    /// the diffuse color for each face.
+    ///
+    /// @param[in] ambient_color,diffuse_color,specular_color the
+    /// default color information for each face.
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/io/writers/SurfaceMeshWriter.h:100]: According to the doxygen doc, you can put several parameters after one @param if they are separated by commas.

File: src/DGtal/shapes/SurfaceMesh.h (hunk 7 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/SurfaceMesh.h:62]: Agreed. I will make a proposition

File: src/DGtal/shapes/SurfaceMesh.h (hunk 10 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/SurfaceMesh.h:120]: do you have a preference (for the one to keep) ?

File: src/DGtal/io/writers/SurfaceMeshWriter.h (hunk 33 above)
[the remark this answers is not in the document]
Comment (dcoeurjo) [reply src/DGtal/io/writers/SurfaceMeshWriter.h:100]: nice

File: src/DGtal/shapes/SurfaceMesh.h (hunk 10 above)
Comment (dcoeurjo) [reply src/DGtal/shapes/SurfaceMesh.h:120]: According to other classes, maybe "SomeTypes" (e.g. Vertices)

File: src/DGtal/shapes/SurfaceMesh.h (hunk 7 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/SurfaceMesh.h:62]: Done

File: src/DGtal/shapes/SurfaceMesh.h (hunk 9 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/SurfaceMesh.h:113]: Done

File: src/DGtal/shapes/SurfaceMesh.h (hunk 10 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/SurfaceMesh.h:120]: Agreed. I also prefer Vertices to VertexRange. The sole problem is when you have a range of Vertices, Verticess is ugly.

Comment (dcoeurjo) [reply src/DGtal/shapes/SurfaceMesh.h:120]: Do you have a type for a range of Vertices in your class?

File: src/DGtal/shapes/SurfaceMesh.h (hunk 12 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/SurfaceMesh.h:247]: done

File: src/DGtal/shapes/SurfaceMesh.h (hunk 14 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/SurfaceMesh.h:625]: Done

File: src/DGtal/shapes/SurfaceMeshHelper.ih (hunk 4 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/SurfaceMeshHelper.ih:72]: done

File: src/DGtal/shapes/doc/moduleSurfaceMesh.dox (hunk 21 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/shapes/doc/moduleSurfaceMesh.dox:29]: done

**Comment by JacquesOlivierLachaud** [comment]
There seems to be an unrelated issue in Travis CI, which requires a modification in Travis CI config. Has Eigen moved to another repository (here bitbucket, but perhaps Eigen has moved to github ?).

**Comment by dcoeurjo** [comment]
@phcerdan has a pending PR for the Eigen URL. On his PR, there is another issue  (timeout on macOS bots) but let me check if I can merge it so that you can benefit from it

**Comment by dcoeurjo** [comment]
@phcerdan ’s PR has just been merged. Should be easier now

**Comment by JacquesOlivierLachaud** [comment]
Cool ! It seems to work now !

**Comment by dcoeurjo** [comment]
Thanks @JacquesOlivierLachaud for the edits!
that's perfect.

**Review by dcoeurjo — APPROVED** [review]

**Pull request merged**